pkg/server/cloud/aws/instances.go

```diff
@@ -359,6 +359,130 @@ func (e *AwsEC2) StartNode(node *api.Node, image cloud.Image, metadata, iamPermi
 	return cloudID, nil
 }
 
+// We need to ensure the dedicated host fulfills two constraints
+// 1) in a state of "available"
+// 2) no tenant is currently occupying the host
+func (e *AwsEC2) ReleaseDedicatedHosts() error {
+	hosts, err := e.listAvailableDedicatedHosts()
+	if err != nil {
+		return err
+	}
+	var hostIdsForRelease []*string
+	for _, host := range hosts {
+		if len(host.Instances) > 0 {
+			continue
+		}
+		hostIdsForRelease = append(hostIdsForRelease, host.HostId)
+	}
+	// TODO do we need to know which hosts were successfully released vs which
+	// were unsuccessful? this runs within the gcLoop so it should clean up
+	// anything not cleaned on a next iteration
+	_, err = e.client.ReleaseHosts(&ec2.ReleaseHostsInput{
+		HostIds: hostIdsForRelease,
+	})
+	return err
+}
+
+func (e *AwsEC2) listAvailableDedicatedHosts() ([]*ec2.Host, error) {
+	describeOutput, err := e.client.DescribeHosts(&ec2.DescribeHostsInput{
+		Filter: []*ec2.Filter{
+			{
+				Name:   aws.String("state"),
+				Values: []*string{aws.String("available")},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return describeOutput.Hosts, nil
+}
+
+func (e *AwsEC2) retrieveOrAllocateHost(node *api.Node) (*string, error) {
+	describeOutput, err := e.client.DescribeHosts(&ec2.DescribeHostsInput{
+		Filter: []*ec2.Filter{
+			{
+				Name:   aws.String("state"),
+				Values: []*string{aws.String("available")},
+			}, {
+				Name:   aws.String("instance-type"),
+				Values: []*string{aws.String(node.Spec.InstanceType)},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	// if there is a host available to accept a new tentant return the host id
+	if len(describeOutput.Hosts) > 0 {
+		return describeOutput.Hosts[0].HostId, nil
+	}
```

Comment:
When `DescribeHosts()` returns available dedicated hosts for the instance type we want to create, does it take into account if it has capacity for the new instance?

Reply:
When you create a dedicated host you have to specify the instance type that the host is for. This will automatically provision the capacity for that instance type. So when we call the `retrieveOrAllocateHost(node *api.Node) (*string, error)` function we are asking it to only return a host of in which suits that instance type and is available to accept a new tenant.